### PR TITLE
Add Edges mode and Matrices output to Mesh Points Scatter node

### PR DESCRIPTION
**Pull request by 3DSinghVFX**
Hi, I have added the *Edges* mode to *Mesh Points Scatter* node that allows scattering random points on edges of the mesh. It also useful in case of *Line(s)* mesh. I have also added the *Matrices* output for random points and these matrices are aligned with normal of the polygons of mesh.

Screenshot from 2020-08-10 22-55-12
 
Screenshot from 2020-08-10 22-54-03

Screenshot from 2020-08-10 22-52-57

### CODE REVIEW

File: animation_nodes/algorithms/mesh/points_scatter.pyx [hunk 1]
```diff
@@ -1,26 +1,38 @@
...
+from ... math cimport (
+    Vector3,
+    addVec3,
+    subVec3,
+    Matrix4,
+    toVector3,
+    distanceVec3,
+    setMatrixTranslation,
+    normalizeVec3_InPlace,
+    )
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/mesh/points_scatter.pyx:25]: Deindent this line.

File: animation_nodes/algorithms/mesh/points_scatter.pyx [hunk 2]
```diff
@@ -89,19 +101,26 @@ cdef LongList computeDistribution(DoubleList triangleWeights, long numberOfPoint
...
     return distribution
 
-cdef Vector3DList sampleRandomPoints(Vector3DList vertices, PolygonIndicesList polygons, LongList distribution,
-                                     long numberOfPoints, long seed):
+cdef Matrix4x4List sampleRandomPoints(Vector3DList vertices, PolygonIndicesList polygons, Vector3DList polyNormals,
+                                      LongList distribution, long numberOfPoints, long seed):
     cdef XoShiRo256Plus rng = XoShiRo256Plus(seed)
-    cdef Vector3DList points = Vector3DList(length = numberOfPoints)
+    cdef Matrix4x4List matrices = Matrix4x4List(length = numberOfPoints)
+    cdef Vector3 guide = toVector3((0, 0, 1))
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/mesh/points_scatter.pyx:108]: Why initialize the variable here?

File: animation_nodes/algorithms/mesh/points_scatter.pyx [hunk 3]
```diff
@@ -89,19 +101,26 @@ cdef LongList computeDistribution(DoubleList triangleWeights, long numberOfPoint
...
-    cdef Vector3 v1, v2, v3, v
     cdef double p1, p2, p3
     index = 0
     for i in range(polygons.getLength()):
         j = i * 3
         v1 = vertices.data[polygons.indices.data[j]]
         v2 = vertices.data[polygons.indices.data[j + 1]]
         v3 = vertices.data[polygons.indices.data[j + 2]]
+
+        direction = polyNormals.data[i]
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/mesh/points_scatter.pyx:119]: Rename this to normal

File: animation_nodes/algorithms/mesh/points_scatter.pyx [hunk 4]
```diff
@@ -113,7 +132,73 @@ cdef Vector3DList sampleRandomPoints(Vector3DList vertices, PolygonIndicesList p
             v.x = p1 * v1.x + p2 * v2.x + p3 * v3.x
             v.y = p1 * v1.y + p2 * v2.y + p3 * v3.y
             v.z = p1 * v1.z + p2 * v2.z + p3 * v3.z
-            points.data[index] = v
+
+            directionToMatrix_LowLevel(matrices.data + index, &direction, &guide, 2, 0)
+            setMatrixTranslation(matrices.data + index, &v)
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/mesh/points_scatter.pyx:137]: Use the `createMatrix` function here. You already have the normal and the tangent (The guide). Normalize them both, compute the bitangent by taking the cross product of the normalized normal and tangent, then pass them to the `createMatrix` function.

File: animation_nodes/algorithms/mesh/points_scatter.pyx [hunk 5]
```diff
@@ -89,19 +101,26 @@ cdef LongList computeDistribution(DoubleList triangleWeights, long numberOfPoint
...
     index = 0
     for i in range(polygons.getLength()):
         j = i * 3
         v1 = vertices.data[polygons.indices.data[j]]
         v2 = vertices.data[polygons.indices.data[j + 1]]
         v3 = vertices.data[polygons.indices.data[j + 2]]
+
+        direction = polyNormals.data[i]
+
+        subVec3(&guide, &v1, &v2)
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/mesh/points_scatter.pyx:121]: Rename guide to tangent.

File: animation_nodes/algorithms/mesh/points_scatter.pyx [hunk 6]
```diff
@@ -113,7 +132,73 @@ cdef Vector3DList sampleRandomPoints(Vector3DList vertices, PolygonIndicesList p
...
+        edgeWeights.data[i] = distanceVec3(vertices.data + i1, vertices.data + i2)
+        edgeWeights.data[i] *= max((weights.get(i1) + weights.get(i2)) / 2.0, 0)
+    return edgeWeights
+
+cdef Matrix4x4List sampleRandomPointsOnEdges(Vector3DList vertices, EdgeIndicesList edges, Vector3DList normals,
+                                             LongList distribution, long numberOfPoints, long seed):
+    cdef XoShiRo256Plus rng = XoShiRo256Plus(seed)
+    cdef Matrix4x4List matrices = Matrix4x4List(length = numberOfPoints)
+    cdef Vector3 direction = toVector3((0, 0, 1))
+    cdef Vector3 guide = toVector3((0, 0, 1))
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/mesh/points_scatter.pyx:171]: Why initialize those here?

File: animation_nodes/algorithms/mesh/points_scatter.pyx [hunk 7]
```diff
@@ -113,7 +132,73 @@ cdef Vector3DList sampleRandomPoints(Vector3DList vertices, PolygonIndicesList p
...
+    cdef Vector3 v
+    cdef Vector3 *v1
+    cdef Vector3 *v2
+    cdef Vector3 *n1
+    cdef Vector3 *n2
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/mesh/points_scatter.pyx:176]: Put those in the same line.

File: animation_nodes/algorithms/mesh/points_scatter.pyx [hunk 8]
```diff
@@ -113,7 +132,73 @@ cdef Vector3DList sampleRandomPoints(Vector3DList vertices, PolygonIndicesList p
...
+        for j in range(distribution.data[i]):
+            p1 = rng.nextDouble()
+            p2 = 1.0 - p1
+
+            v.x = p1 * v1[0].x + p2 * v2[0].x
+            v.y = p1 * v1[0].y + p2 * v2[0].y
+            v.z = p1 * v1[0].z + p2 * v2[0].z
+
+            directionToMatrix_LowLevel(matrices.data + index, &direction, &guide, 0, 2)
+            setMatrixTranslation(matrices.data + index, &v)
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/mesh/points_scatter.pyx:201]: Same thing as above.

File: animation_nodes/nodes/mesh/mesh_points_scatter.py [hunk 9]
```diff
@@ -29,31 +38,46 @@ def create(self):
...
-        return randomPointsScatter(vertices, polygons, weights, seed, max(amount, 0))
+        if self.mode == "POLYGONS":
+            if len(polygons) == 0: return Matrix4x4List()
+            if polygons.polyLengths.getMaxValue() > 3:
+                if self.methodType:
+                    polygons = triangulatePolygonsUsingEarClipMethod(vertices, polygons)
+                else:
+                    polygons = triangulatePolygonsUsingFanSpanMethod(polygons)
+            polyNormals = calculatePolygonNormals(vertices, polygons)
+            return randomPointsScatter(vertices, polygons, polyNormals, weights, seed, max(amount, 0))
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/mesh_points_scatter.py:76]: To unify things, rename the function to `scatterPointsOnPolygons` and rename the other function to `scatterPointsOnEdges`.

File: animation_nodes/nodes/mesh/mesh_points_scatter.py [hunk 10]
```diff
@@ -4,17 +4,26 @@
...
+]
+
 class MeshPointsScatterNode(bpy.types.Node, AnimationNode):
     bl_idname = "an_MeshPointsScatterNode"
     bl_label = "Mesh Points Scatter"
 
+    mode: EnumProperty(name = "Mode", default = "POLYGONS",
+        items = modeItems, update = AnimationNode.refresh)
+
     methodType: BoolProperty(name = "Use Advanced Method for Mesh sampling", default = False,
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/mesh_points_scatter.py:27]: While at it, rename this to `useAdvancedTriangulationMethod` and mention triangulation in the name.

File: CHANGELOG.md [hunk 11]
```diff
@@ -37,6 +37,8 @@
 - Added mesh *Triangulation* methods and *Triangulate Mesh* node.
 - Added Mesh Points Scatter node.
 - Added support for 2D vectors in 3D viewer node.
+- Added matrices output in *Mesh Points Scatter* node.
+- Added edges mode in *Mesh Points Scatter* node.
```
Comment (OmarEmaraDev) [inline CHANGELOG.md:41]: Change log entries now goes to the Unreleased section above.

File: animation_nodes/algorithms/mesh/points_scatter.pyx [hunk 12]
```diff
@@ -113,7 +132,73 @@ cdef Vector3DList sampleRandomPoints(Vector3DList vertices, PolygonIndicesList p
...
-            points.data[index] = v
+
+            directionToMatrix_LowLevel(matrices.data + index, &direction, &guide, 2, 0)
+            setMatrixTranslation(matrices.data + index, &v)
+            index += 1
+
+    return matrices
+
+
+def randomPointsScatterForEdges(Vector3DList vertices, EdgeIndicesList edges, Vector3DList normals,
```
Comment (OmarEmaraDev) [inline animation_nodes/algorithms/mesh/points_scatter.pyx:143]: Rename normals to `vertexNormals`

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 2 above)
Comment (3DSinghVFX) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:108]: I thought initializing `toVector3((0, 0, 1))` in loop may affect the performance.

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 4 above)
Comment (3DSinghVFX) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:137]: Omar, should I call that function from `mesh/c_utils.pyx` or define it here?

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 6 above)
Comment (3DSinghVFX) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:171]: Removed.

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 8 above)
Comment (3DSinghVFX) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:201]: Done.

File: animation_nodes/nodes/mesh/mesh_points_scatter.py (hunk 9 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/mesh_points_scatter.py:76]: Done.

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 3 above)
Comment (3DSinghVFX) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:119]: Done.

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 2 above)
Comment (3DSinghVFX) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:108]: Removed.

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 1 above)
Comment (3DSinghVFX) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:25]: Done.

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 7 above)
Comment (3DSinghVFX) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:176]: Done.

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 12 above)
Comment (3DSinghVFX) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:143]: Done.

File: animation_nodes/nodes/mesh/mesh_points_scatter.py (hunk 10 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/mesh_points_scatter.py:27]: Is it correct `useAdvancedTriangulationMethod: BoolProperty(name = "Use Ear Clip Triangulation Method",
                                    default = False, update = propertyChanged)` ?

File: CHANGELOG.md (hunk 11 above)
Comment (3DSinghVFX) [reply CHANGELOG.md:41]: Done.

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 4 above)
Comment (OmarEmaraDev) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:137]: Move it somewhere common like `math/matrix`.

File: animation_nodes/nodes/mesh/mesh_points_scatter.py (hunk 10 above)
Comment (OmarEmaraDev) [reply animation_nodes/nodes/mesh/mesh_points_scatter.py:27]: Yes, looks good.

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 4 above)
Comment (3DSinghVFX) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:137]: Okay.

File: animation_nodes/nodes/mesh/mesh_points_scatter.py (hunk 10 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/mesh_points_scatter.py:27]: Okay.

File: animation_nodes/algorithms/mesh/points_scatter.pyx (hunk 4 above)
Comment (OmarEmaraDev) [reply animation_nodes/algorithms/mesh/points_scatter.pyx:137]: Also, perhaps name it something more approperiate like `matrixFromNormalizedAxisData`.

**Comment by 3DSinghVFX** [comment]
@OmarEmaraDev For the matrix, I have aligned the Z-axis along the vertex/polygon normal.

File: animation_nodes/math/matrix.pyx [hunk 13]
```diff
@@ -316,3 +316,10 @@ cdef float getMatrix3x3PartDeterminant(Matrix3_or_Matrix4 *m):
         m.a12 * m.a21 * m.a33 -
         m.a11 * m.a23 * m.a32
     )
+
+cdef void createMatrix(Matrix4 *m, Vector3 *center, Vector3 *tangent,
```
Comment (OmarEmaraDev) [inline animation_nodes/math/matrix.pyx:320]: Rename this to `matrixFromNormalizedAxisData`. Also in the pxd file.

Comment (3DSinghVFX) [reply animation_nodes/math/matrix.pyx:320]: Okay.

Comment (3DSinghVFX) [reply animation_nodes/math/matrix.pyx]: Done.

**Pull request merged**